pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -317,3 +334,23 @@ func (nm *NodeManager) GetVSphereInstance(nodeName k8stypes.NodeName) (VSphereIn
 	}
 	return *vsphereInstance, nil
 }
+
+// renewNodeInfo renews vSphere connection, VirtualMachine and Datacenter for NodeInfo instance.
+func (nm *NodeManager) renewNodeInfo(nodeInfo *NodeInfo, reconnect bool) (*NodeInfo, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vsphereInstance := nm.vsphereInstanceMap[nodeInfo.vcServer]
+	if vsphereInstance == nil {
+		err := fmt.Errorf("vSphereInstance for vSphere %q not found while refershing NodeInfo for VM %q", nodeInfo.vcServer, nodeInfo.vm)
+		return nil, err
+	}
+	if reconnect {
+		err := vsphereInstance.conn.Connect(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	vm := nodeInfo.vm.RenewVM(vsphereInstance.conn.GoVmomiClient)
```

Comment:
It looks like you have created a new VM object and new datacenter object and replaced the old nodeinfo. Then do you need to release/destroy the old VM and datacenter object? There is a DeleteVM method in virtualmachine.go. I am wondering if replace current NodeInfo with new objects will leave dangling objects behind.

Reply:
@luomiao the DeleteVM function does deletion of actual vSphere VM. We don't want to do that. Our intention here is just to renew the vclib VirtualMachine struct, which in turn extends govmomi VirtualMachine, by setting up new client object. The vSphere VM should remain intact. 

Any on going operation that still has handle to older vclib VirtualMachine object can continue to use it without any foreseeable race condition (for now at least :)). Once those operations finish the older VirtualMachine struct instance, which is no longer being referenced by NodeInfo or any other ongoing operation, will get garbage collected. 

Hope that answers your concern.

Reply:
Got it, thanks for the explanation :)